plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -79,97 +82,124 @@ Then, in the _body_, we track the term and type application (`Apply` or `TyInst`
-[Identifying fully applied call sites].
-
-Because a function can be called in the `body` multiple times and may not be fully applied for all
-its calls, we cannot simply keep a simple substitution map like in `UnconditionalInline`,
-which substitute *all* occurrences of a variable.
-
-Below are some more examples:
-
-Example 1: function in body
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_.
 
-let f = \x . x
-in let g = \y . f
-   in g a
+Note that over-application of a function would also be inlined. An example of over-application:
 
-`f` and `g` each has 1 lambda. However, `g`'s _body_ includes `f` which also has a lambda.
-Since we only count the number of lambdas, `g` is fully applied, and we inline.
-`g a` reduces to `f`, which reduces the amount of code. Again, this also opens up more dead code
-elimination opportunities.
+```haskell
+let id = \y -> y
+     f = \x -> id
+ in f x y
+```
 
-Example 2: function as an argument
+`f`'s arity is 1, but is applied to 2 arguments. We inline `f` in this case if its cost and size
+are acceptable.
 
-let id :: forall a . a -> a
-    id x = x
-in id {int -> int} (\x -> x +1)
-
-Here we have a type application for a function that takes one type variable.
-I.e., it's fully applied in terms of type.
-In terms of terms, `id` takes one argument, and is indeed applied to one.
-So `id` is fully applied! And we inline it.
-Inlining and reducing `id` reduces the amount of code, as desired.
-
-Example 3: function application in _rhs_
-
-let f = (\x.\y.x+y) 4
-in f 5
-
-With beta-reduction, `f` becomes `\y.4+y` and it has 1 lambda.
-The _body_ `f 5` is a fully applied function!
-We can reduce it to 4+5.
-
-Example 4: applications and lambda abstractions in _body_
-
-let f = \x.\y.x+y
-in (\z.f 3 4) 2
-
-With beta-reduction, the _body_ becomes `f 3 4` and thus `f` is fully applied.
+Because a function can be called in the `body` multiple times and may not be fully applied for all
+its calls, we cannot simply keep a substitution map like in `Inline`,
+which substitute *all* occurrences of a variable. We store the function in a map
+`Utils.CalledVarEnv` with all information needed for inlining saturated functions.
 
 (2) How do we decide whether cost and size are acceptable?
 
 We currently reuse the heuristics 'Utils.sizeIsAcceptable' and 'Utils.costIsAcceptable'
-that are used in 'UnconditionalInline'. For
+that are used in unconditional inlining. For
 
 let v = \x1.\x2...\xn.VBody in body
 
 we check `VBody` with the above "acceptable" functions.
-Note that all `LamAbs` and `TyAbs` should have been
+Note that all `LamAbs` and `TyAbs` are not acceptable in terms of size, but they should have been
 counted out already so we should not immediately encounter those in `VBody`.
 Also, we currently reject `Constant` (has acceptable cost but not acceptable size).
 We may want to check their sizes instead of just rejecting them.
```

Comment:
I thought it makes sense to discuss some of that here because the reader may wonder why we chose the heuristics - I'm saying we haven't much yet, we'll fine-tune it. Maybe I should make that more clear.